storage/innobase/os/os0file.cc

```diff
@@ -2340,7 +2330,11 @@ AIO::is_linux_native_aio_supported()
 	}
 
 	ut_free(buf);
-	close(fd);
+	if (!srv_read_only_mode) {
+		my_close(fd, MYF(0));
+	} else {
+		close(fd);
+	}
```

Comment:
I see, the `fd` was created either by `innobase_mysql_tmpfile()` or `open("ib_logfile0", O_RDONLY|O_CLOEXEC)`. Maybe we could always use the same file here? The `innobase_mysql_tmpfile()` looks quite strange, because the intention seems to be to test whether `io_submit()` can be used on the InnoDB data and redo log files.

Anyway, that could be the topic of another pull request. I was just wondering why we need two variants of `close()`.

Reply:
as the tmpfile could be on a different filesystem to ib_logfile0 I suspect both are needed.

Reply:
This function was renamed from `os_aio_native_aio_supported()` in mysql/mysql-server@98909cefbc37e54efc6452c7e95bccbf64ac9213. The use of tmpfile was introduced much earlier, in mysql/mysql-server@ad494212e8d3e209a9eb1f623686ea37e89a1051, which is referring to WL#6445, which is not public. Apparently the purpose of that work was to introduce the parameter `innodb_read_only` (`srv_read_only_mode`).

This check looks bogus to me, but cleaning it up does not need to be part of this pull request.

Later, MySQL 5.7 introduced true InnoDB temporary tables. It is true that it could make sense to check if the shared temporary tablespace supports `io_submit()`. But, if I remember correctly, we would not allow the creation of InnoDB temporary tables in `innodb_read_only` mode. And, even if we are read-only, I’d expect to be able to test `io_submit()` with some non-destructive operation, such as `IOCB_CMD_PREAD` or `IOCB_CMD_NOOP`.